### PR TITLE
feat: vendor CI log groups

### PR DESCRIPTION
**Pull request by cymruu**
### Description
Adds support for grouping tasks logs for some CI vendors. Closes https://github.com/vercel/turbo/issues/6188

### Testing Instructions
I've added some unit tests, sadly I need guidance how to test it more reliably as the changes were made in rust code which is not currently used in generated turbo binary.

### CODE REVIEW

**Comment by chris-olszewski** [comment]
I'll work on getting this reviewed and ready for merge, but it probably won't get shipped until the Rust codepath is feature complete and relatively bug free. (Only 3 minor features to finish up and then roughly 40 integration tests to fixup) From a first glance this looks great. Thanks!

**Comment by arlyon** [comment]
Hey small update here, we are in the process of ripping out the go code in the next week or two at which point we have free reign to merge this in. I'll update here when that is done :)

**Comment by cymruu** [comment]
@arlyon I'm happy to hear that. Thanks for heads up!

File: crates/turborepo-lib/src/task_graph/visitor.rs [hunk 1]
```diff
@@ -367,13 +378,16 @@ impl<'a> Visitor<'a> {
...
-        if self.run_opts.is_github_actions {
-            let package = if self.run_opts.single_package {
-                None
+        if let Some(vendor_behavior) = vendor_behavior {
+            let group_name = if self.run_opts.single_package {
+                ""
             } else {
-                Some(task_id.package())
+                task_id.package()
             };
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/task_graph/visitor.rs:386]: We to include both package name and task name in the group name if it's a multi-package repo and the task name if it is just single package mode. This matches previous behavior.

e.g.
```
let group_name = if self.run_opts.single_package {
    task_id.task().to_string()
} else {
    format!("{}:{}", task_id.package(), task_id.task())
}
```

File: crates/turborepo-ci/src/vendors.rs [hunk 2]
```diff
@@ -243,6 +266,10 @@ pub(crate) fn get_vendors() -> &'static [Vendor] {
                     sha_env_var: Some("GITHUB_SHA"),
                     branch_env_var: Some("GITHUB_REF_NAME"),
                     username_env_var: Some("GITHUB_ACTOR"),
+                    behavior: Some(VendorBehavior {
+                        group_prefix: |group_name, _| format!("::group::{group_name}"),
+                        group_suffix: |_, _| String::from("::endgroup::"),
```
Comment (chris-olszewski) [inline crates/turborepo-ci/src/vendors.rs:271]:
```suggestion
                        group_prefix: |group_name, _| format!("::group::{group_name}\n"),
                        group_suffix: |_, _| String::from("::endgroup::\n"),
```

File: crates/turborepo-lib/src/task_graph/visitor.rs [hunk 3]
```diff
@@ -367,13 +378,16 @@ impl<'a> Visitor<'a> {
...
-            let (header, footer) = github_header_footer(package, task_id.task());
+            let (header, footer) = (
+                (vendor_behavior.group_prefix)(group_name, task_start_time),
+                (vendor_behavior.group_suffix)(group_name, task_start_time),
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/task_graph/visitor.rs:389]: We'll want to defer computing the footer until later as we won't know until the task finishes what the correct task end time will be.

Given that this constraint complicates the design a bit, I think it's reasonable to either:
 - Leave this as-is with a TODO comment explaining that the timestamp will be incorrect
 - Remove the Gitlab CI vendor prefix/suffix implementation as it is the only one that uses the timestamp.

**Review by chris-olszewski — CHANGES_REQUESTED** [review]
First off, thank you so much for your patience and this great PR.

I think the design is very solid and only have some minor behavior suggestions.


As far as testing. The only test we have that currently is [log-order-github.t](https://github.com/vercel/turbo/blob/e834b541bfefdc714fcf4d36bc16fe44b43f5291/turborepo-tests/integration/tests/run-logging/log-order-github.t)

You can run this with the following commands:
```
cd turborepo-tests/integration
python3 -m venv .cram_env
python3 -m pip install --upgrade pip
pip install "prysk==0.15.2"
.cram_env/bin/prysk --shell=bash tests/run-logging/log-order-github.t
```

I don't think we need additional testing of each individual vendor output. If you do want to verify that another vendor's CI header/footer, you can copy that Github test and update it with the appropriate env var and logs.

File: crates/turborepo-lib/src/task_graph/visitor.rs (hunk 3 above)
Comment (cymruu) [reply crates/turborepo-lib/src/task_graph/visitor.rs]: I decided to remove the GitlabCI vendor for now. I'll be interested in adding it again in a separate pull request

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 4175321eab38658908b8c240e2299d9481b576fb

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Pull request merged**